Add grpc reflection to grpc_json_transcoder.

hq6 opened the pull request:
Previously, the grpc_json_transcoder filter required a statically
generated set of protobuf descriptors to perform translation.

This commit adds the ability to fetch protobuf descriptors via grpc
reflection.

Note that, per discussion on #1182, this change entails a fairly major
restructuring of the filter's code.

In addition to unit tests, this PR was tested manually using `greeter_server` from grpc examples in the grpc repository.

Signed-off-by: Henry Qin <root@hq6.me>

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/config.cc:
```diff
@@ -28,8 +29,49 @@ GrpcJsonTranscoderFilterConfig::createRouteSpecificFilterConfigTyped(
+
+    AsyncReflectionFetcherSharedPtr async_reflection_fetcher =
+        std::make_shared<AsyncReflectionFetcher>(proto_config.reflection_cluster_config(),
+                                                 proto_config.services(), async_clients,
```
async_clients should be implementation details of AsyncReflectionFetcher.  It should be inside the class

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/async_reflection_fetcher.h:
```diff
@@ -0,0 +1,157 @@
+   * onReceiveMessage callback, since the same object is used for handling all
+   * the callbacks.
+   */
+  Envoy::Thread::MutexBasicLockable receive_message_mutex_;
```
Since you are using envoy grpc,  you don''t need thread protection,   its callbacks are called in the same thread when the requests are started.  They will be all in the envoy main thread.

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/BUILD:
```diff
@@ -14,10 +14,20 @@ envoy_extension_package()
+    hdrs = [
+        "async_reflection_fetcher.h",
+        "async_reflection_fetcher_callbacks.h",
+        "descriptor_pool_builder.h",
```
making two separate libraries  for descriptor_pool_builder and asyn_reflection_fetcher

hq6 replied:
Will do.

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/async_reflection_fetcher.cc:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  // Record that we have already received the response for this Rpc.
+  remaining_callbacks_.erase(callbacks);
```
need to handle error case,   maybe move line 80 at top of the function?

qiwzhang replied:
I am fine with the policy of if fetching fail, not to ready envoy.    But we need to make the error bold and clear, not require users to turn on debug log to debug it.   We could either crash the envoy by throw exception, or log error.

qiwzhang replied:
I see.  logging error is fine here for parsing the response.   But what about fetching failed cases, did we log errors?

hq6 replied:
This following is the only failing fetch case I'm aware of. Are there others that I am not?

```
  void onRemoteClose(Envoy::Grpc::Status::GrpcStatus status, const std::string& message) override {
    if (status != Envoy::Grpc::Status::Ok) {
      ENVOY_LOG(error, "transcoding_filter: server reflection request failed with status {}: {}",
                status, message);
    }
  }
```

qiwzhang replied:
This should be fine then.

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/descriptor_pool_builder.cc:
```diff
@@ -0,0 +1,156 @@
+            api_->fileSystem().fileReadToEnd(proto_config_.proto_descriptor()))) {
+      throw EnvoyException("transcoding_filter: Unable to parse proto descriptor");
+    }
+    loadFileDescriptorProtos(descriptor_set);
```
Do we need to sort dependency in this case.  maybe we can call descriptor_pool_available directly?

qiwzhang replied:
If the sorting is not needed,  please don't do it.   Even it is on the filter loading time in the main thread, it will delay the envoy to be ready, it will use memory and cpu.

hq6 replied:
Will separate and remove the extra sorting.

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/descriptor_pool_builder.cc:
```diff
@@ -0,0 +1,156 @@
+    }
+  }
+
+  descriptor_pool_available_(std::move(descriptor_pool));
```
maybe call "on_descriptor_pool_ready"?

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/descriptor_pool_builder.cc:
```diff
@@ -0,0 +1,156 @@
+      DescriptorSetCase::kReflectionClusterConfig:
+    // The constructor of this object is responsible for ensuring the async_reflection_fetcher_ is
+    // not null.
+    async_reflection_fetcher_->requestFileDescriptors(
```
Ideally,  async_reflection_fetcher should be created here, as it is implementation detail of DescriptorPoolBuilder.   For easy unit-test to pass in a mock async_fetcher,  you can use a asyn_fetch_factory, so it can be mocked.

qiwzhang replied:
number of lines doesn't matter here,  it is where the logic belongs matter.    A module design with clear logic in separate modules is much better than codes lump all logic into one function.

hq6 replied:
For the record, I believe this is a matter of software design philosophy, but I always conform to local style, so I'll update this since it sounds like you feel strongly about it.

hq6 reviewed (commented):
Thank you for the review!

Please see comments inline and let me know if 
a) you want to insist on any of the points I've expressed a different opinion on, or 
b) whether you're willing to approve if I don't change those parts of this PR

I will hold off on modifying this PR further until you've expressed an opinion in either direction, so as to avoid wasted work until we've reached agreement on the target structure of the code.

qiwzhang commented:
Open source code, at lease Envoy project,  requires code quality. It is NOT me that insisting on the way I want to do,  It is the code quality. 

There is a big problem that I have commented and you did not reply,  the freeing of the callbacks.   Let me try to find that comment or re-comment it again.

qiwzhang commented on source/extensions/filters/http/grpc_json_transcoder/async_reflection_fetcher.cc:
```diff
@@ -0,0 +1,93 @@
+    Envoy::Grpc::AsyncClient<ServerReflectionRequest, ServerReflectionResponse>& async_client =
+        async_clients_[i];
+
+    AsyncReflectionFetcherCallbacks* callbacks = new AsyncReflectionFetcherCallbacks(
```
Where is the `callbacks` freed? ???

hq6 replied:
Sorry I missed this earlier.

This is a good catch, that they are not freed!

However, it does leave an interesting question:
When do you think it's safe to free them in case of failure?
For success, I think we can free them after we set init_target_->ready().
For failure, there are approaches I can currently think of:
a. We can keep a count of any kind of return, success or failure, and free when that count is reached.
b. We can record that an error happened, and erase from `remaining_callbacks_` just like in success. 
When `remaining_callbacks_` reach zero, we release regardless of success or failure, but only call ready() when the error flag is not set.

Thoughts on which is preferable, or do you have other ideas for deciding when to free?

qiwzhang replied:
Actually, you do have `callbacks_` to hold the unique_ptr for them,  they are freed eventually when `AsyncReflectionFetcher` is freed.

Just a suggestion,  `remaining_callbacks_` can be just a integer to keep track of number of outstanding requests.

hq6 replied:
I made it a set because I was not sure if internal retries in the Envoy gRPC library could cause duplicate responses.

Do you happen to know if it duplicate responses are guaranteed to be handled and de-duped?

qiwzhang replied:
All envoy async callbacks are guaranteed to be called once.  Not need to worry about been called twice.  In your code, if onCloseStream() is called on network failure,  it will not call your callback.  But that is fine since we decide not to "ready" envoy in failed cases.

qiwzhang commented:
BTW,  I like to see separate unit-test xyz_test.cc for the two new classes,  ProtoDescriptorBuilder and AsyncReflectionFetcher.   If they have dependency, mock their dependency.  

Thanks

adisuissa commented on api/envoy/extensions/filters/http/grpc_json_transcoder/v3/transcoder.proto:
```diff
@@ -103,6 +103,17 @@ message GrpcJsonTranscoder {
   }
 
+  message ReflectionConfig {
+    // Supplies the cluster name of the gRPC cluster to use for reflection.
+    // This will typically be the same gRPC cluster that Envoy is forwarding
+    // gRPC traffic to, but it does not need to be.
```
Is the gRPC cluster that Envoy forwards to is the default value for this field? If so, please explicitly note so. If not, what is the default value?

hq6 replied:
I don't think there is a default value for this field in the PR as written (beyond the protobuf default).  
This was an oversight on my part, thanks for catching!
Based on [this post](https://github.com/envoyproxy/envoy/issues/12074), it looks like Envoy can forward to multiple gRPC clusters.

So I have three questions.
1.Do you think we should specify a default value?
2. If we do specify a default value and there are multiple gRPC clusters configured, should we pick one arbitrarily or throw an error or log an error and do not initialize?
3. If we do not specify a default value, do you have thoughts on which of the following is preferable?
a) Logging an error and not initializing when no value is specified.
b) Throwing an exception and crashing Envoy when no value is specified.

adisuissa replied:
To follow up on the 3rd option you suggested, in the API we have ability to denote a string field as mandatory by setting its length ([example](https://github.com/envoyproxy/envoy/blob/0202852eb0d1423f86f72cb5ef735d459ffd9183/api/envoy/config/cluster/v3/cluster.proto#L807)).
The config will be rejected in case this is not populated.
If the value of this field needs to be declared somewhere else, please clarify the docs accordingly.

qiwzhang replied:
How about using `envoy::config::core::v3::GrpcService`?    `cluster_name` is only for envoy_grpc.  GrpcService can support other grpc.

But on the other hand, this grpc server is the upstream server, normal grpc requests should be routed by envoy upstream.   So it is nature to use envoy_grpc.

hq6 replied:
I'll defer to @adisuissa here, as the API shepherd.

adisuissa commented on api/envoy/extensions/filters/http/grpc_json_transcoder/v3/transcoder.proto:
```diff
@@ -103,6 +103,17 @@ message GrpcJsonTranscoder {
+
+    // The amount of time in ms that Envoy should wait for gRPC reflection
+    // requests to return.
+    uint64 request_timeout = 2;
```
Could this be of type `google.protobuf.Duration`?
What is the default value?

hq6 replied:
Yes, I can make it  a Duration. I'll document the default value regardless.

qiwzhang replied:
How about adding `config.core.v3.RetryPolicy retry_policy`.  Since network failures are so bad,  it is better to have retry.  Can add it in a later pr.

qiwzhang replied:
For `request_timeout`,  you can do

```
google.protobuf.Duration timeout = 3 [(validate.rules).duration = {
    required: true
    gte {}
  }];
```

hq6 replied:
It's a fair point. I'll add it in this PR if it's easy.

adisuissa commented on changelogs/current.yaml:
```diff
@@ -21,6 +21,9 @@ behavior_changes:
 
 minor_behavior_changes:
 # *Changes that may cause incompatibilities for some users, but should not for most*
+- area: grpc_json_transcoder
+  change: |
+    Support grpc reflection for building json transcoding.
```
Is this a new feature or a modified behavior?

hq6 replied:
It depends on the Envoy project's definition of a feature, so I'm not sure.
If I had to guess, I would guess it's a new feature because it enables a new capability that didn't exist before.

adisuissa replied:
In this case this update should probably be under a different category.

qiwzhang replied:
It is a new feature

hq6 replied:
Thanks for confirming! Will move.

adisuissa commented on docs/root/configuration/http/http_filters/grpc_json_transcoder_filter.rst:
```diff
@@ -104,3 +104,18 @@ gRPC or RESTful JSON requests to localhost:51051.
+Modified Envoy configuration to use gRPC reflection
+---------------------------------------------------
+Rather than using a statically generated configuration, Envoy can fetch service
+definitions via grpc reflection. This can be done by replacing
+``proto_descriptor`` in the filter's config with the following:
```
Does this imply that the `proto_descriptor` and `reflection_cluster_config` are mutually exclusive?
if so this should be clarified in the API (proto comments).

hq6 replied:
Yes. I will make sure to make it clear in the comments.

adisuissa reviewed (commented):
Thanks!
Left a few API comments.

hq6 commented:
> Open source code, at lease Envoy project, requires code quality. It is NOT me that insisting on the way I want to do, It is the code quality.
> 

For the record, I disagree with the Envoy project's definition of code quality. 
I believe understandability is more important than modularity, and subscribe to John Ousterhout's philosophy of deep abstractions to reduce complexity.

However, I will update the code to conform for consistency's sake.

> There is a big problem that I have commented and you did not reply, the freeing of the callbacks. Let me try to find that comment or re-comment it again.

Thanks for finding it again and re-commenting. I also did not find the original one, maybe GitHub lost it?

hq6 commented:
> BTW, I like to see separate unit-test xyz_test.cc for the two new classes, ProtoDescriptorBuilder and AsyncReflectionFetcher. If they have dependency, mock their dependency.
> 
> Thanks

Yes, I think you pointed this out in your comments on my original draft PR, so I separated it out in this one.
If there's something else I didn't separate out that you asked for, please remind!

hq6 commented:
@qiwzhang @adisuissa It'll probably take me quite a while to address all the review comments. 
Should I close this PR or make it a draft or keep it open in the interim?

adisuissa commented:
> @qiwzhang @adisuissa It'll probably take me quite a while to address all the review comments, since some of the changes @qiwzhang requested are structural (Note that this is a personal project so I have limited time to work on it).
> 
> Should I close this PR or make it a draft or keep it open in the interim while I address the comments?

As long as the PR keeps the same spirit, and is not completely different, then we can keep it open in "waiting" state for a while. This will keep the previous comments.
If you think it is going to be significantly different, then you can open a new PR.

/wait

hq6 commented:
@qiwzhang I am still waiting on your opinion on [this question](https://github.com/envoyproxy/envoy/pull/23061#discussion_r970354609) before this work on this PR can proceed further.

I agree with you that it is an important problem, as you mentioned.

qiwzhang commented:
@hq6 I replied it.  Thanks.

qiwzhang commented:
@hq6 Overall this pr is good. Thanks for the hard work.  It just needs a bit of restructure in classes `ProtoDescriptorBuilder` and `AsyncReflectionFetcher` with proper unit-tests.   I don't see a big changes from this pr, so we can continue to use this pr.

I can sense some defensive tone in your comment.  If my comment somehow offended you, I am sorry, it was not my intention. 

Anyway,  this is a good feature to have.  Thanks for the hard work.

hq6 commented:
> I can sense some defensive tone in your comment. If my comment somehow offended you, I am sorry, it was not my intention.

No offense taken, although I do object to any claims that there's one globally accepted definition of code quality today. 
I currently believe that a majority of software developers today hold as a basic tenet that modularity is more important than ease of understanding, and I am part of the minority that disagrees.

I simply wanted to make it clear that this contribution to Envoy does not mean I endorse Envoy's school of thought about what code quality means. :)

route33 commented:
Is this merge stuck? when is going to be released?

hq6 commented:
This PR cannot be merged until comments are addressed, and code coverage is up to the required level. 
I've been busy with my day job, and haven't yet had time to address the comments.

jeffsawatzky commented:
@hq6 are you still working on this? Or has this been cancelled?

hq6 commented:
@jeffsawatzky I'm not actively working on this due to other commitments.
I was planning to try to get to it after tax season, but if someone else wants to take over the PR by addressing the remaining comments and fixing code coverage, I am not opposed.

For coordination reasons, it'd be good for such a person to comment on this PR or the related issue to avoid duplicated work.

the-duke commented:
@hq6 awesome work keep it up. This would be an extremely useful feature!
